src/test/app/Credentials_test.cpp

```diff
@@ -0,0 +1,782 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2012, 2013 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <test/jtx.h>
+#include <xrpld/app/tx/detail/Credentials.h>
+#include <xrpld/ledger/ApplyViewImpl.h>
+#include <xrpl/basics/strHex.h>
+#include <xrpl/protocol/Feature.h>
+#include <xrpl/protocol/Indexes.h>
+#include <xrpl/protocol/Protocol.h>
+#include <xrpl/protocol/TxFlags.h>
+#include <xrpl/protocol/jss.h>
+
+#include <iostream>
+#include <string_view>
+
+namespace ripple {
+namespace test {
+
+// Helper function that returns the owner count of an account root.
+static inline std::uint32_t
+ownerCnt(test::jtx::Env const& env, test::jtx::Account const& acct)
+{
+    std::uint32_t ret{0};
+    if (auto const sleAcct = env.le(acct))
+        ret = sleAcct->at(sfOwnerCount);
+    return ret;
+}
+
+static inline bool
+checkVL(
+    std::shared_ptr<SLE const> const& sle,
+    SField const& field,
+    std::string const& expected)
+{
+    auto const b = sle->getFieldVL(field);
```

Comment:
b is for debug, I'll remove it
hex because  `sle->getFieldVL(field)` return blob, which can have 0 or other non-printable characters, to be sure that string compared correctly. It is a test anyway.